### ISSUE TITLE
Packages.conf try_install no longer installs any packages on systems that use symlinked resolv.conf in live environment

### ISSUE THREAD

**Issue by ArrayBolt3** [calamares/calamares#1942]
While chasing down a Lubuntu 22.04 bug related to LibreOffice localization, I discovered that none of the packages listed in Lubuntu's packages.conf `try_install` section were being installed. Further debugging revealed that the APT database in the newly installed system was not being updated, despite the presence of `update_db: true` in the packages.conf file.

After doing a lot of testing, I finally tracked down the issue to a regression caused by this commit: https://github.com/calamares/calamares/pull/1692/commits/a888156debbe7a5f0db33c5ac2a457e668fe8283

Because Lubuntu 22.04 is using systemd-resolved to handle network configuration, /etc/resolv.conf is a symlink within the live system. Due to the commit mentioned above, Calamares does not copy the resolv.conf file from the live environment into the chroot. As a result, APT is unable to get internet access during the installation, resulting in the inability to update the package database or install any packages.

By manually editing /lib/x86_64-linux-gnu/calamares/modules/networkcfg/main.py, I removed the `follow_symlinks=False` bits from the file and then did another test installation of Lubuntu. APT was able to update the database and install the listed packages with this modification.

FWIW, I am testing this in a QEMU/KVM virtual machine created with virt-manager on Ubuntu Studio 21.10. Firmware is BIOS (not EFI), architecture is amd64.

Steps to reproduce the issue:

1: Download the Lubuntu 22.04 ISO from https://lubuntu.me/downloads/
2: Spin up a virtual machine using the new ISO. I set the OS to "Ubuntu 21.04" (since that's the newest OS virt-manager can recognize on my system), and gave the VM 4 GB RAM, 2 CPU cores, and 20 GB of disk space.
3: Once the live environment has booted, open /etc/calamares/modules/packages.conf, and change the try_install section so that it looks like this:

    - try_install
      - emacs

Save your changes and exit.
4: Install Lubuntu 22.04, accepting all the default settings. I used the "Erase disk" option for partitioning, and set my name to "User", username as "user", password as "z", and enabled autologin.
5: When the installation is complete and you're at the "Finish" screen, uncheck "Restart now" and click "Done".
6: Mount the disk you installed Lubuntu onto, to /mnt (I did `sudo mount /dev/vda1 /mnt`).
7: Bind mount /dev, /sys, /proc, /run, and /run/udev to /mnt/dev, /mnt/sys, /mnt/proc, /mnt/run, and /mnt/run/udev, respectively.
8: Run `sudo chroot /mnt`.
9: Type "emacs".

Expected results: Emacs should launch.

Actual results: command-not-found informs me that emacs is not installed.

Steps to work around the issue:

1: After step 3 in the issue reproduction steps, open /lib/x86_64-linux-gnu/calamares/modules/networkcfg/main.py, and edit the file to reverse the commit mentioned in the description. Save your changes and exit
2: Complete the rest of the issue reproduction steps.

Result: Emacs launches.

My idea for fixing the problem would be to detect if the installed system's resolv.conf is a symlink or not. Either way, the installer **should** overwrite the installed system's resolv.conf with the file present in the live environment. If the file is a symlink, the installer should make a note of the path pointed to by the symlink before overwriting it. When the installer is finishing up, it should delete the copied resolv.conf and replace it with the original symlink if the file was a symlink to begin with. (Perhaps this could be implemented as a "networkcfgfinal" module that distros could insert immediately before the umount module?) This should avoid bringing back bug https://github.com/calamares/calamares/issues/1630, while still allowing the packages module to install packages into the newly installed system.

**Comment by monetaryabyss** [comment]
Since the fix for this appears to be applied in recent versions of Calamares, I think this issue should be closed as a downstream packaging issue in (L)Ubuntu.

**Comment by demmm** [comment]
Make sure to always post which version of Clamares is in use, so there is no needlessly trying to chase bugs already fixed upstream.  Have you tried using the latest Calamares and can still reproduce this bug?
https://calamares.io/issues/

**Comment by kkofler** [comment]
Well, the default `mount.conf` does not bind-mount `/run`, it just mounts a blank `tmpfs` there in the chroot. (Or has that changed recently?)

**Comment by adriaandegroot** [comment]
That was going to be my next, and final, point: the configuration mounts a *tmpfs* on the target root, rather than a *bind* mount, so that's why the copied symlink doesn't point anywhere useful. The fix for this, downstream -- but possibly also in the example -- is to change `mount.conf` to bind-mount `/run`, just like it already does `/run/udev`.

**Comment by kkofler** [comment]
I think it makes more sense to fix `/etc/nsswitch.conf`.

**Comment by ArrayBolt3** [comment]
@adriaandegroot Yes, I do know about the Calamares log, I simply didn't bother extracting it from the VM since that's a cumbersome task and there was enough data to reproduce the issue without it. Sorry, I will know better in the future and will include the install log should I need to file a bug report.

Thank you so much for your help! Your solution with bind-mounting `/run` does indeed appear to work, and is likely to be the easiest way to fix the problem. I'll see if I can get that implemented. (I really did think that this was a Calamares issue since my code mod appeared to work... Thank you for helping me figure out what was actually going wrong!)

**Comment by kkofler** [comment]
I wrote:
> I think it makes more sense to fix `/etc/nsswitch.conf`.

To elaborate on that: See: https://blogs.gnome.org/mcatanzaro/2020/12/17/understanding-systemd-resolved-split-dns-and-vpn-configuration/

> A Word about Ubuntu
>
> Although Ubuntu has used systemd-resolved for four years now, it has _not_ switched from nss-dns to nss-resolve, contrary to upstream recommendations. This means that on Ubuntu, glibc still reads `/etc/resolv.conf`, finds 127.0.0.53 listed there, and then makes an IP connection to systemd-resolved rather than talking to it via varlink or D-Bus, as occurs on Fedora. The practical effect is that, on Ubuntu, you can still manually edit /etc/resolv.conf and applications will respond to those changes, unlike Fedora. Of course, that would be a disaster, since it would cause all of your DNS configuration in systemd-resolved to be completely ignored. But it’s still possible on Ubuntu. On Fedora, that won’t work at all.
>
> If you’re using custom VPN software that doesn’t work with systemd-resolved, chances are it probably tries to write to `/etc/resolv.conf`.

That was posted in 2020, more than a year ago. Why does Ubuntu still insist on shipping this broken configuration, even in the light of issues like this one (the Calamares issue I am replying to)? And why can this not be changed at least on the images that use Calamares? Why do you think it is safer to bind-mount `/run`, which can cause all sorts of issues because your chroot environment will see all files from the host (not just `resolv.conf`) and the other way round?

**Comment by ArrayBolt3** [comment]
@kkofler Not sure why they do that. Initially, I wasn't able to figure out how to even do that (couldn't find the right man page turns out), but I just found some documentation that I believe explains what I was doing wrong, so I can test your solution, too. I'll ask around and find out what's going on.

I don't know for sure that it's safer to bind-mount `/run`, and I actually was intending to do some testing that would verify if it was safe or not. However, reading your comment, I'm reallizing that, if all I need is one file from out of systemd-resolved, I can probably bind-mount just the one directory that has the file I need and have a way safer fix. In the event that the `/etc/nsswitch.conf` fix isn't viable immediately, this would be a solution that would be way more safe than exposing all of `/run` to the newly installed system. Thanks for saying something, you probably just saved me a lot of testing work!
